lib/charms/mysql/v0/s3_helpers.py

```diff
@@ -0,0 +1,157 @@
+        logger.exception(
+            f"Failed to upload content to S3 bucket={s3_parameters['bucket']}, path={content_path}",
+            exc_info=e,
+        )
+        return False
+
+    return True
+
+
+def list_backups_in_s3_path(s3_parameters: Dict) -> List[str]:
+    """Retrieve subdirectories in an S3 path.
+
+    Args:
+        s3_parameters: A dictionary containing the S3 parameters
+            The following are expected keys in the dictionary: bucket, path,
+            region, endpoint, access-key and secret-key
+
+    Returns: a list of subdirectories directly after the S3 path.
+
+    Raises: any exception raised by boto3
+    """
+    try:
+        logger.info(
+            f"Listing subdirectories from S3 bucket={s3_parameters['bucket']}, path={s3_parameters['path']}"
+        )
+        s3_client = boto3.client(
+            "s3",
+            aws_access_key_id=s3_parameters["access-key"],
+            aws_secret_access_key=s3_parameters["secret-key"],
+            endpoint_url=s3_parameters["endpoint"],
+            region_name=s3_parameters["region"],
+        )
+        list_objects_v2_paginator = s3_client.get_paginator("list_objects_v2")
+        s3_path_directory = (
+            s3_parameters["path"]
+            if s3_parameters["path"][-1] == "/"
+            else f"{s3_parameters['path']}/"
+        )
+
+        directories = []
+        for page in list_objects_v2_paginator.paginate(
+            Bucket=s3_parameters["bucket"],
+            Prefix=s3_path_directory,
+            Delimiter="/",
+        ):
+            for content in page.get("Contents", []):
+                key = content["Key"]
+                if ".md5" in key:
+                    directories.append(
+                        key.lstrip(s3_path_directory).split("/")[0].split(".md5")[0]
+                    )
+
+        return directories
+    except Exception as e:
+        logger.exception(
+            f"Failed to list subdirectories in S3 bucket={s3_parameters['bucket']}, path={s3_parameters['path']}",
+            exc_info=e,
+        )
+        raise
+
+
+def fetch_and_check_existence_of_s3_path(s3_parameters: Dict, path: str) -> bool:
+    """Checks the existence of a provided S3 path by fetching the object.
+
+    Args:
+        s3_parameters: A dictionary containing the S3 parameters
+            The following are expected keys in the dictionary: bucket, region,
+            endpoint, access-key and secret-key
+        path: The path to check the existence of
+
+    Returns: a boolean indicating the existence of the s3 path
+
+    Raises: any exceptions raised by boto3
+    """
+    s3_client = boto3.client(
+        "s3",
+        aws_access_key_id=s3_parameters["access-key"],
+        aws_secret_access_key=s3_parameters["secret-key"],
+        endpoint_url=s3_parameters["endpoint"],
+        region_name=s3_parameters["region"],
+    )
+
+    try:
+        response = s3_client.get_object(Bucket=s3_parameters["bucket"], Key=path)
```

Comment:
if we check that file exists, why we need to download it all?
```suggestion
        response = s3_client.get_object(Bucket=s3_parameters["bucket"], Key=path, Range="0-1")
```

can be handled in separate PR